src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java

```diff
@@ -740,8 +751,20 @@ public void close() {
     UploadCommand upload = new UploadCommand(uri, casUpload, key, wrappedIn, length);
     Channel ch = null;
     boolean success = false;
+    File compressedUpload = null;
     if (storedBlobs.putIfAbsent((casUpload ? CAS_PREFIX : AC_PREFIX) + key, true) == null) {
       try {
+        if (casUpload && this.compressCasUploads) {
+          // first compress to a temporary file before creating an upload channel
+          compressedUpload = createCompressedUpload(key, in);
+          in.close();
+          in = new FileInputStream(compressedUpload);
+          InputStream wrappedCompressedIn = new FilterInputStream(in) {
+            @Override
+            public void close() {}
+          };
+          upload = new UploadCommand(uri, true, key, wrappedCompressedIn, compressedUpload.length());
+        }
```

Comment:
Can we move this up and wrap only once? Something on the lines of
```java
InputStream input;
if (casUpload && this.compressCasUploads) {
    compressedUpload = createCompressedUpload(key, in);
    input = new FileInputStream(compressedUpload);
else {
    input = in;
}
InputStream wrappedIn =
    new FilterInputStream(input) { 
```

Reply:
👍